proposals/p1964.md

```diff
@@ -0,0 +1,356 @@
+but are fundamentally designed for the same purpose. Languages that have a
+dedicated character data type generally include character literals, for example
+C++, Java, Swift to name a few. Whereas other languages that lack distinct
+character type, like Python use strings of length one to serve the same purpose
+a character data type. For more information see
+[Character Literals Wiki](https://en.wikipedia.org/wiki/Character_literal),
+[Character Literals DBpedia](https://dbpedia.org/page/Character_literal)
+
+## Proposal
+
+Put character literals in single quotes, like `'a'`. Character literals work
+like numeric literals:
+
+-   Every different literal value has its own type.
+-   The bit width is determined by the type of the variable the literal is
+    assigned to, not the literal itself. Follows the plan from #1934.
+-   A character literal will model single Unicode code points that have a single
+    concrete numerical representation. We will not be supporting other
+    formulations like code unit sequences or grampheme clusters as these will be
+    modeled with normal string literals.
+
+## Details
+
+-   A character literal is a sequence enclosed with single quotes delimiter ('),
+    of UTF-8 code units that must be a valid encoding. This matches
+    [the UTF-8 encoding of Carbon source files](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0142.md#character-encoding).
+-   A character literal must encode exactly one code point.
+-   It supports addition and subtraction. These operations produce another code
+    point literal, or produce an error if the result is out of range.
+-   Character literals support some back-slash (`\`) escape sequences, including
+    `\t`, `\n`, `\r`, `\"`, `\'`, `\\`, `\0`, and `\u{HHHH...}`. See
+    [String Literals: Escape sequence](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0199.md#escape-sequences).
+-   Character literals implicitly convert to Unicode strings, even though the
+    numeric values of the encoding may change.
+
+We will not support:
+
+-   character literals that don't contain exactly one Unicode code point;
+-   multi-line literals;
+-   "raw" literals (using #'x'#);
+-   `\x` escape sequences;
+-   character literals with a single quote (`'`) or back-slash (`\`), except as
+    part of an escape sequence;
+-   empty character literals (`''`);
+-   ASCII control codes (0...31), including whitespace characters other than
+    word space (tab, line feed, carriage return, form feed, and vertical tab),
+    except when specified with an escape sequence.
+
+### Types
+
+We will have the types `Char8`, `Char16`, and `Char32` representing code units
+in UTF-8, UTF-16, and UTF-32, but we will not support all code units, but only
+those which map directly to the complete value of a code point. However,
+character literals will use their own types distinct from these:
```

Comment:
There is a tension here that you would need a large type like `Char32` to hold any code point, but it seems like we are going to allow using smaller types to hold smaller code points, so we can support things like `var c: Char8 = 'a';`.

Reply:
Thinking about this further:

- I think it makes sense and will have applications to be able to store a code point in a `Char8`, `Char16`, or `Char32`, as long as it fits in the specified number of bits. People are going to at least want to 
- I don't think it makes sense to store code units in `u8`, `u16`, or a `u32`, since we don't want all those operations and wrapping behavior.
- I think it is reasonable to either introduce a new type for code units or to use the `CharN` types. If I am not mistaken, code points are a strict subset of code units, which causes me to think we should use the `CharN` types for both.

Reply:
I agree we should use `CharN` types for both, but that means they are really code *unit* types.

I still think we should revisit this when we're actually fully building out encoding, decoding, etc. Right now, it seems hard to really even see the complete picture for me.

But I think there is utility in a type that can carry any runtime code *point*. I'm not especially worried about efficiency here, folks who need that would likely better operate on the `CharN` types once they are available.

I view this is more analogous to something like a `BigInt` type and its relationship with integer literals, with the handy benefit of us being able to trivially define the limits on the possible values it needs to represent.

Reply:
I don't think there is a need for a `BigInt`-like type for code points. I believe all code points fit in a `Char32` and in fact 32-bit code points are the same as the 32-bit code units.

Reply:
This makes sense, thank you! For now I have created a new section for future work for UTF encoding types, and updated the type section to use `CharN` for both code points and code units